File: src/optic/lua/asm/StepVisitor.java
package optic.lua.asm;

import java.util.*;

public interface StepVisitor<T, X extends Throwable> {
	T visitComment(String comment) throws X;

	T visitAssignment(Register register, RValue value) throws X;

	T visitBlock(AsmBlock block) throws X;

	T visitBreakIf(RValue condition, boolean isTrue) throws X;

	T visitDeclaration(VariableInfo variable) throws X;

	T visitForEachLoop(List<VariableInfo> variables, RValue iterator, AsmBlock body) throws X;

	T visitForRangeLoop(VariableInfo counter, RValue from, RValue to, RValue step, AsmBlock body) throws X;

	T visitIfElseChain(LinkedHashMap<FlatExpr, AsmBlock> clauses) throws X;

	T visitLoop(AsmBlock body) throws X;

	T visitReturn(List<RValue> values) throws X;

	T visitSelect(Register out, int n, RValue varargs) throws X;

	T visitVoid(RValue.Invocation invocation) throws X;

	T visitWrite(VariableInfo target, RValue value) throws X;

	default List<T> visitAll(List<Step> values) throws X {
		var result = new ArrayList<T>(values.size());
		for (var v : values) {
			result.add(v.accept(this));
		}
		return result;
	}
}

File: src/optic/lua/asm/StepVisitor.kt
package optic.lua.asm

import optic.lua.asm.instructions.*
import optic.lua.messages.CompilationFailure
import org.jetbrains.annotations.Contract
import java.util.*

abstract class StepVisitor<R> {
    abstract fun defaultValue(x: Step): R

    @Throws(CompilationFailure::class)
    open fun visitBlock(x: Block): R {
        return defaultValue(x)
    }

    @Throws(CompilationFailure::class)
    open fun visitIfElseChain(x: IfElseChain): R {
        return defaultValue(x)
    }

    @Throws(CompilationFailure::class)
    open fun visitComment(x: Comment): R {
        return defaultValue(x)
    }

    @Throws(CompilationFailure::class)
    open fun visitDeclare(x: Declare): R {
        return defaultValue(x)
    }

    @Throws(CompilationFailure::class)
    open fun visitForRangeLoop(x: ForRangeLoop): R {
        return defaultValue(x)
    }

    @Throws(CompilationFailure::class)
    open fun visitGetVarargs(x: GetVarargs): R {
        return defaultValue(x)
    }

    @Throws(CompilationFailure::class)
    open fun visitRead(x: Read): R {
        return defaultValue(x)
    }

    @Throws(CompilationFailure::class)
    open fun visitReturn(x: Return): R {
        return defaultValue(x)
    }

    @Throws(CompilationFailure::class)
    open fun visitSelect(x: Select): R {
        return defaultValue(x)
    }

    @Throws(CompilationFailure::class)
    open fun visitToNumber(x: ToNumber): R {
        return defaultValue(x)
    }

    @Throws(CompilationFailure::class)
    open fun visitWrite(x: Write): R {
        return defaultValue(x)
    }

    @Throws(CompilationFailure::class)
    open fun visitInvoke(x: Invoke): R {
        return defaultValue(x)
    }

    @Throws(CompilationFailure::class)
    open fun visitAssign(x: Assign): R {
        return defaultValue(x)
    }

    @Throws(CompilationFailure::class)
    fun visitAll(steps: List<Step>): List<R> {
        val list = ArrayList<R>(steps.size)
        for (step in steps) {
            list.add(visit(step))
        }
        return list
    }

    @Throws(CompilationFailure::class)
    fun visit(x: Step): R {
        return when (x) {
            is Block -> visitBlock(x)
            is IfElseChain -> visitIfElseChain(x)
            is Comment -> visitComment(x)
            is Declare -> visitDeclare(x)
            is ForRangeLoop -> visitForRangeLoop(x)
            is GetVarargs -> visitGetVarargs(x)
            is Read -> visitRead(x)
            is Return -> visitReturn(x)
            is Select -> visitSelect(x)
            is ToNumber -> visitToNumber(x)
            is Write -> visitWrite(x)
            is Invoke -> visitInvoke(x)
            is Assign -> visitAssign(x)
            else -> throw IllegalArgumentException(x.javaClass.name)
        }
    }

}
